File: src/Java/algorithms/datastructure/SimplifyPath.java
package algorithms.datastructure; 

import java.util.*; 

public class SimplifyPath {
    /* Function: Given a Unix absolute path with /..//, return the simple one. 
     * Solution: Split by '/' and push to stack. If . or empty, continue; If 
     *     .., pop. If stack is empty, return /. 
     */ 
    
    public String simplifyPath(String path) {
        if (path == null || path.length() == 0) {
            return ""; 
        }
        
        String[] paths = path.split("/"); 
        Stack<String> stack = new Stack<String>(); 
        for (String cur : paths) {
            if (cur.equals(".") || cur.length() == 0) {
                continue; 
            }
            
            if (cur.equals("..")) {
                if (!stack.empty()) {
                    stack.pop(); 
                }
            } else {
                stack.push(cur); 
            }
        }
        
        StringBuffer sb = new StringBuffer(); 
        while (!stack.empty()) {
            sb.insert(0, "/" + stack.pop()); 
        }
        
        if (sb.length() == 0) {
            return "/"; 
        }
        return sb.toString(); 
    }
}

File: src/CPP/src/leetcode/SimplifyPath.cpp
/*
 * [Source] https://leetcode.com/problems/simplify-path/
 * [Difficulty]: Medium
 * [Tag]: Stack
 * [Tag]: String
 */

#include <iostream>

using namespace std;

// [Solution]: Split string by '/'. Use a stack. If find "..", pop, if "." or empty, do nothing.
// [Corner Case]: only "/", then stack is empty.
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/datastructure/SimplifyPath.java
 */

/* Java solution
public class Solution {
    public String simplifyPath(String path) {
		if(path==null || path.length()<=0) return null;
		if(path.equals("/")) return "/";
		char[] array=path.toCharArray();
        Stack<String> stack=new Stack<String>();
		int idx=1;
		while(idx<array.length){
			if(array[idx]=='/'){ // a slash
				idx++;
				continue;
			}
			String dir=getDirectory(array, idx);
			if(!dir.equals(".")){ // current dir
				if(dir.equals("..")){ // parent dir
					if(!stack.empty()) stack.pop();
				}else{
					stack.push(dir);
				}
				idx=idx+dir.length()+1;
			}else{
				idx=idx+2;
			}
		}
		if(stack.empty()) return "/";
		Stack<String> reverse=new Stack<String>();
		while(!stack.empty()){
			reverse.push(stack.pop());
		}
		StringBuffer result=new StringBuffer();
		while(!reverse.empty()){
			result.append('/');
			result.append(reverse.pop());
		}
		return result.toString();
    }
	
	public String getDirectory(char[] path, int start){
		StringBuffer directory=new StringBuffer();
		int index=start;
		while(index<path.length){
			if(path[index]=='/') break;
			directory.append(path[index]);
			index++;
		}
		return directory.toString();
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
